traversal/procedure/GraphProcedure.java

```diff
@@ -181,39 +184,39 @@ public Builder() {
             this.vertices = new HashMap<>();
         }
 
-        // TODO: remove when the query planner orders vertices natively
-        public GraphProcedure orderAndbuild() {
-            return new GraphProcedure(orderedVertices());
-        }
-
         public GraphProcedure build() {
-            return new GraphProcedure(vertices.values().stream().sorted(Comparator.comparing(ProcedureVertex::order))
+            return new GraphProcedure(vertices.values().stream().sorted(comparing(ProcedureVertex::order))
                     .toArray(ProcedureVertex[]::new));
         }
 
-        /**
-         * TODO: this is a compatibility layer to translate edge-ordered procedures into vertex-ordered procedures
-         * we need to remove it once the query planner natively orders vertices instead of edges
-         */
-        private ProcedureVertex<?, ?>[] orderedVertices() {
-            ProcedureVertex<?, ?>[] orderedVertices = new ProcedureVertex[vertices.size()];
-            List<ProcedureVertex<?, ?>> vertexList = iterate(vertices.values()).toList();
-            vertexList.sort(Comparator.comparing(v -> v.lastInEdge() == null ? 0 : v.lastInEdge().order()));
-            for (int i = 0; i < vertexList.size(); i++) {
-                orderedVertices[i] = vertexList.get(i);
-                orderedVertices[i].setOrder(i);
+        private void streamlineVertices() {
+            Stack<ProcedureVertex<?, ?>> open = new Stack<>();
+            vertices.values().stream().filter(ProcedureVertex::isStartingVertex).forEach(open::add);
+            Set<ProcedureVertex<?, ?>> closed = new HashSet<>();
+
+            int vertexOrder = 0;
+            while (!open.empty()) {
+                ProcedureVertex<?, ?> vertex = open.pop();
+                vertex.setOrder(vertexOrder++);
+                closed.add(vertex);
+                vertex.outs().stream().map(ProcedureEdge::to).distinct().sorted(comparing(v -> -v.ins().size())).filter(
+                        v -> v.ins().stream().map(ProcedureEdge::from).allMatch(closed::contains)
+                ).forEach(open::add);
             }
-            return orderedVertices;
+            assert vertexOrder == vertices.size();
```

Comment:
The optimiser ultimately only cares about the directions of the edges in the plan. `0 -> 2 -> 4 <- 3 <- 1` and `0 -> 1 -> 4 <- 3 <- 2` are therefore equivalent plans, which lead to inefficient traversal. 
This procedure is meant to rectify that by essentially performing a depth-first topological sort on the vertices. For the top vertex in the `open` stack, we put on the `open` stack that part of its neighbourhood that already has its requirements fulfilled, i.e. all the immediately preceding vertices are already present in the `closed` set. These newly considered vertices are sorted such that the vertex with the greatest amount of incoming edges ends up on the top, so that during traversal we can start backtracking earlier from the intersection vertices without having to consider the other branches.
This algorithm can be improved, but it is a simple and reasonably effective first approximation.